### ISSUE TITLE
[Feature]: Change all cgo interface with arrow

### ISSUE THREAD

**Issue by xiaofan-luan** [milvus-io/milvus#16733]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe.

currently, milvus use protobuf heavily in the cgo interface for data exchange between go and c code.

However, if we take a deep look into it, many of the code path is copied and converted heavily, for instance in the load path:
1. read parquet buffer from minio (GO)
2. use arrow parquet reader to read arrow from parquet buffer(Go)
3. convert arrow to protobuf and sent to cpp via cgo
4. parse protobuf, convert to memory format in segcore 

### Describe the solution you'd like.

My suggestion is that we change all interface  between go and c should use apache arrow and that's what arrow designed for.

We should reduce memory consumption, accelerate search/load speed once we done that.

Further step is to convert internal data structure in segcore and knowhere into cgo but we still have no idea about whether this is doable.

**Comment by xiaofan-luan** [comment]
anyone interest pleas assign yourself

**Comment by jiaoew1991** [comment]
I will try this one

**Comment by jiaoew1991** [comment]
Hi @xiaofan-luan , I am working on this issue, the scope seams quite large. So I start from "Delete" function in "segment_c.h" for a POC. 

When I work on this issue, I found some blockers, and the way I handle it is very dirty:

1.  the "internal/core" module does not depends on Arrow, but the "internal/storage" module depends on it, so I change the compile order, make "storage" compile first then let "core" depends on the arrow compiled with "storage"
2. for unknown reasons, I cannot compile correctly, I've been tortured by this for days, after lots of attempts, When I upgraded rocksdb it works fine  ┓( ´∀` )┏, so I upgrade the rocksdb from 6.x to 7.1 . 

I want to raise another question, does milvus really need to depend on rocksdb? I found another component [pebble](https://github.com/cockroachdb/pebble) maintained by cockroachdb which written by pure Go, it seems easier to use 😄
